stl/inc/expected

```diff
@@ -685,6 +689,303 @@ public:
         }
     }
 
+    template <class _Uty = _Err>
+    _NODISCARD constexpr _Err error_or(_Uty&& _Other) const& noexcept(
+        is_nothrow_copy_constructible_v<_Err>&& is_nothrow_convertible_v<_Uty, _Err>) { // strengthened
+        static_assert(
+            is_copy_constructible_v<_Err>, "is_copy_constructible_v<E> must be true. (N4928 [expected.object.obs]/20)");
+        static_assert(
+            is_convertible_v<_Uty, _Err>, "is_convertible_v<U, E> must be true. (N4928 [expected.object.obs]/20)");
+
+        if (_Has_value) {
+            return _STD forward<_Uty>(_Other);
+        } else {
+            return _Unexpected;
+        }
+    }
+
+    template <class _Uty = _Err>
+    _NODISCARD constexpr _Err error_or(_Uty&& _Other) && noexcept(
+        is_nothrow_move_constructible_v<_Err>&& is_nothrow_convertible_v<_Uty, _Err>) { // strengthened
+        static_assert(
+            is_move_constructible_v<_Err>, "is_move_constructible_v<E> must be true. (N4928 [expected.object.obs]/22)");
+        static_assert(
+            is_convertible_v<_Uty, _Err>, "is_convertible_v<U, E> must be true. (N4928 [expected.object.obs]/22)");
+
+        if (_Has_value) {
+            return _STD forward<_Uty>(_Other);
+        } else {
+            return _STD move(_Unexpected);
+        }
+    }
+
+    // [expected.object.monadic]
+
+    template <class _Fn>
+        requires is_copy_constructible_v<_Err>
+    constexpr auto and_then(_Fn&& _Func) & {
+        using _Uty = remove_cvref_t<invoke_result_t<_Fn, _Ty&>>;
+
+        static_assert(_Is_specialization_v<_Uty, expected>,
+            "expected<T,E>::and_then(F) requires the return type of F to be a specialization of expected "
+            "N4928 [expected.object.monadic]/3.");
+
+        if (this->_Has_value) {

```

Comment:
is `this` required?

Reply:
It's not: I copied parts of the `std::optional` implementation for consistency and the members are stored in dependent base classes there, so `this->` is required. I can remove these ones if we like.

Reply:
I have no preferences on `this`, but code style conventions for this library to use `this->` only when it required; see https://github.com/microsoft/STL/pull/2267#discussion_r751827069

Reply:
We conventionally avoid unnecessary `this->`-qualification to make the necessary uses stand out.